seccomp/tracer.go

```diff
@@ -205,45 +235,77 @@ func (t *syscallTracer) seccompSessionAdd(s seccompSession) error {
 }
 
 func (t *syscallTracer) seccompSessionDelete(pid uint32) {
+	var closeFds []int32
 
 	t.seccompSessionMu.Lock()
 
-	s, ok := t.seccompSessionMap[pid]
+	fd, ok := t.seccompSessionMap[pid]
 	if !ok {
 		t.seccompSessionMu.Unlock()
 		logrus.Warnf("Unexpected error: seccomp fd not found for pid %d", pid)
 		return
 	}
 
-	// From the moment in which we delete a given seccompSession from the
-	// sessions map, no new read-operation can be launched from the tracer
-	// for this very fd being eliminated. That's to say that the order of
-	// instructions in this block is important: "delete" first, StopWait()
-	// second.
 	delete(t.seccompSessionMap, pid)
-	t.seccompSessionMu.Unlock()
 
-	// Alert the tracer's pollServer to stop polling on this seccomp-fd.
-	if err := t.pollsrv.StopWait(s.fd); err != nil {
-		logrus.Errorf("failed StopWait execution of seccomp fd %d for pid %d: %v",
-			s.fd, pid, err)
-	}
+	if t.service.closeSeccompOnContExit {
+		cntrID, ok := t.pidToContMap[pid]
+
+		if ok {
+			var cntrInitPid uint32
+
+			cntr := t.service.css.ContainerLookupById(cntrID)
+			if cntr != nil {
+				cntrInitPid = cntr.InitPid()
+			}
+
+			// if the container is no longer there, or the pid being deleted is the
+			// container's init pid, we close all seccomp fds for that container
+			if cntr == nil || pid == cntrInitPid {
+				closeFds = t.seccompSessionCMap[cntrID]
+				delete(t.seccompSessionCMap, cntrID)
+			}
+
+			delete(t.pidToContMap, pid)
+		} else {
+			logrus.Warnf("Unexpected error: pid %d is not associated with a container", pid)
+		}
 
-	// Now we are finally ready to close the seccomp-fd. We don't want to do
-	// this any earlier as kernel could potentially re-assign the same fd to
-	// new seccomp-bpf sessions without giving us a chance to complete the fd
-	// unregistration process.
-	if err := syscall.Close(int(s.fd)); err != nil {
-		logrus.Errorf("failed to close seccomp fd %v for pid %d: %v", s.fd, pid, err)
+	} else {
+		closeFds = []int32{fd}
 	}
 
-	logrus.Debugf("Removed session for seccomp-tracee: %v", s)
+	t.seccompSessionMu.Unlock()
+
+	if len(closeFds) > 0 {
```

Comment:
Why are you adding this 'if' statement? Looks like isn't needed in this logic.

Reply:
I added it because it's possible for []closeFds to be empty, and I wanted the debug message in L301 to come out only when we closed at least one FD.

Reply:
got it